### PR TITLE
make DHT routing table refresh interval configurable.

### PR DESCRIPTION
**Pull request by thomas-yuan**
As I have mentioned in #129, I am testing how to reduce bandwidth usage without performance lost. 
I changed refresh interval to 30 seconds, seems it works fine. So I think maybe we can make the refresh interval also configurable?

Here is the result on 4 computers. (github support upload image now. :+1: )
- Default means enable 'readonly', all other settings are default.
- Tuning means enable 'readonly'(#128), disable re-bootstrap(#162), using 'ping' instead of 'get_peers' (#155) and set refresh interval to 30 seconds (this PR :smile:)

There is one exception in performance test, if removed that one, then get_immutable_item average response will be received in 200 ~ 300 ms. ( :+1: )

performance ms
performance ms 2
routingtable

### CODE REVIEW

**Comment by arvidn** [comment]
I'm a little bit worried about increasing this timeout. I'm not sure your measurements are all that relevant. I think the risk with increasing this is that the nodes you have in your table go stale. Just knowing about them doesn't mean they're not stale. As soon as you realized they are stale, obviously you'll replace them, and the counter will stay high, bit it doesn't reflect how long the stale node was in there.

basically, how long did it take to discover that it was stale?

**Comment by thomas-yuan** [comment]
[Here] (https://github.com/thomas-yuan/libtorrent/tree/dht_perf) is a test branch. Basically I modified dht_put.cpp to accept different settings, and send get_immutable_item every 60 minutes (that's where the performance statistics comes from.) and run for 24 hours.

Based on my test, the average number of live nodes in routing table is ~330, if the refresh interval is 5 seconds, then we need about 27 minutes to refresh all nodes.  If change it to 30, then 27 * 6, about 2 hours.

My thought is, 
- this change is just make the DHT more configurable, the default value will be same as before. It gives an option when user try to run it on mobile device where data usages is a concern. (Tbh, I even thought I should test 60 seconds)
- reduce max_fail_count when increase refresh interval maybe help. didn't test yet. (Now it's 20, maybe change to 3?)
- user's actions on DHT can help refresh table. 
- If the node runs without "readonly" feature, other nodes can help itself to refresh nodes.

File: include/libtorrent/session_settings.hpp [hunk 1]
```diff
@@ -1503,6 +1504,10 @@ namespace libtorrent
 		// the number of seconds a immutable/mutable item will be expired.
 		// default is 0, means never expires.
 		int item_lifetime;
+
+		// the number of seconds DHT routing table need to be refreshed, to
+		// maintain enough fresh nodes in routing table.
+		int refresh_timeout;
```
Comment (arvidn) [inline include/libtorrent/session_settings.hpp]: now that I think about it. The original paper talks about refresh intervals in terms of how frequently a single node should be pinged, which iirc, was one hour. The 5 second counter here really is derived from 1 hour / number_of_nodes. It's an optimization to even out the cost of the routing table maintenance. I wonder if it would make more sense for the setting to be specified in the term of how often a single node is pinged.

File: src/kademlia/dht_tracker.cpp [hunk 2]
```diff
@@ -197,7 +197,7 @@ namespace libtorrent { namespace dht
 		m_blocker.set_rate_limit(m_settings.block_ratelimit);
 
 		error_code ec;
-		m_refresh_timer.expires_from_now(seconds(5), ec);
+		m_refresh_timer.expires_from_now(seconds(m_settings.refresh_timeout), ec);
```
Comment (arvidn) [inline src/kademlia/dht_tracker.cpp]: if so, this would be calculated by dividing the timeout by the number of nodes in the routing table (probably with some lower bound, like 50-100 nodes). It's important that the actual timeout doesn't get too long, because it's likely that the number of nodes will increase while we're waiting

**Comment by thomas-yuan** [comment]
I closed this PR because I can't prove the routing table will be good enough.

**Comment by arvidn** [comment]
I'm not opposed to merging this, but I think the setting should be specified in the ping interval for a single node rather than the delta. It really should be adjusted by the size of the routing table.
